### PR TITLE
Add tensorflow op documentation.

### PR DESCRIPTION
**Pull request by dsantl**
Tensorflow Op documentation.

### CODE REVIEW

File: user_technical/reference_guide/tensorflow.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+### API
+Memgraph Op API consists of inputs, attributes and outputs.
+
+#### Input
+There are two  inputs:
+* query
+* input list
+
+The Query is a string. The query is *openCypher* query supported by Memgraph. Memgraph Op has some limitations on the query. Limitations are explained [here](#limitations).
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: - is -> is an
- *openCypher* -> `openCypher`

File: user_technical/reference_guide/tensorflow.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    memgraph_op_module.memgraph_op( \
+    query_holder, \
+    input_list_holder, \
+    output_dtype = tf.int64)
+      
+  # Run tensorflow graph and give input to tensorflow graph
+  sess.run(memgraph_op, {query_holder: query, input_list_holder: input_list})
+```
+
+Query execution replaces ```$input_list``` with provided op input ([see python example for more](#python-example)). Input list is the only **query parameter** used by Memgraph Op.
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: I wouldn't bold the "query parameter".

File: user_technical/reference_guide/tensorflow.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+#### Attributes
+
+Memgraph Op attributes:
+* host, default = 127.0.0.1
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: I suggest this format: 

* `host` (default: `127.0.0.1`)

File: user_technical/reference_guide/tensorflow.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+* port, default = 7687
+* user, default = "" (empty string)
+* password = "" (empty string)
+* use_ssl = false
+* output_dtype
+
+*Host*, *port*, *user*, *password* and *use_ssl* are attributes used
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: *text* -> `text`

File: user_technical/reference_guide/tensorflow.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+Memgraph Op attributes:
+* host, default = 127.0.0.1
+* port, default = 7687
+* user, default = "" (empty string)
+* password = "" (empty string)
+* use_ssl = false
+* output_dtype
+
+*Host*, *port*, *user*, *password* and *use_ssl* are attributes used 
+for connecting to Memgraph. The only different attribute is *output_type*.
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: type or dtype?

File: user_technical/reference_guide/tensorflow.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+* password = "" (empty string)
+* use_ssl = false
+* output_dtype
+
+*Host*, *port*, *user*, *password* and *use_ssl* are attributes used 
+for connecting to Memgraph. The only different attribute is *output_type*. 
+*output_dtype* has no default value and 
+it is used to determine the type of output tensor. 
+Notice that all data in the output tensor must be of the same type.
+*dtype* can be **int64**, **double**, **bool** and **string**.
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: **text** -> `text`

File: user_technical/reference_guide/tensorflow.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+#### Outputs
+
+Memgraph Op has two outputs:
+* header
+* rows
+
+The header is a string list. The list contains headers provided by query execution:
+
+```openCypher
+MATCH (n) return n.name as Name, n.address as Address;
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: as -> AS
return -> RETURN

File: user_technical/reference_guide/tensorflow.md [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+```
+
+Header is ["Name", "Address"].
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: put the list in `monospaced` font

File: user_technical/reference_guide/tensorflow.md (hunk 2 above)
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: Use max 80 characters per line.

File: user_technical/reference_guide/tensorflow.md [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+Rows data represent query result. Rows data is the matrix (rows x headers). If there are no results from the query (empty set), the matrix has a dimension (0 x 0).
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: represent -> represents the
is the -> is a matrix with dimensions `(|rows| x |headers|)`

File: user_technical/reference_guide/tensorflow.md [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+
+##### Using lists as part of the output
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: Title Case

File: user_technical/reference_guide/tensorflow.md [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+
+Let's see following example:
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: see -> see the

File: user_technical/reference_guide/tensorflow.md [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+
+```openCypher
+MATCH (n)-->(m) RETURN n.id as id, collect(m.value) as value_list;
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: as -> AS

File: user_technical/reference_guide/tensorflow.md [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+```
+
+This query returns *n.id* and list of *m.values*. Memgraph Op returns a matrix. Therefore all elements in the matrix must be of the same type. Memgraph Op expands lists into the row. Matrix dimension is rows x (standard headers + sum of list sizes).
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: *italic* -> `monospaced`
add some kind of formatting around the math expression

File: user_technical/reference_guide/tensorflow.md [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+|3  |[8,8,8,1,2]|
+
+Headers: 
+
+|id|value_list_0|value_list_1|value_list_2|
+|---|---|---|---|
+
+Matrix output:
+
+|||||||
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: why empty first row?

File: user_technical/reference_guide/tensorflow.md [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+|---|---|---|---|---|---|
+|1|1|2|3|4|5|
+|2|5|4|8|1|2|
+|3|8|8|8|1|2|
+
+Memgrap Op supports also more than one list in the output:
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: supports also -> also supports

File: user_technical/reference_guide/tensorflow.md [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+
+```openCypher
+MATCH (n)-->(m) RETURN n.id as id, collect(m.value) as value_list, collect(m.id) as neigh;
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: as -> AS
collect -> COLLECT

File: user_technical/reference_guide/tensorflow.md [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+```
+
+Query output:
+
+|id |value_list|neigh|
+|---|---       |---|
+|1  |[1,2,3,4,5]|[1, 2]|
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: inconsistent formatting.

File: user_technical/reference_guide/tensorflow.md [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+|||||||
+|---|---|---|---|---|---|
+|1|1|2|3|4|5|1|2|
+|2|5|4|8|1|2|3|4|
+|3|8|8|8|1|2|1|3|
+
+### Limitations
+
+#### Input list
+Input list (```$input_list```) can contain only elements of **int64** type.
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: **bold** -> `monospaced`

File: user_technical/reference_guide/tensorflow.md [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+
+#### Output types:
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: No punctuations in the title

File: user_technical/reference_guide/tensorflow.md [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+Output matrix contains only elements with the same data type. The data type can be **int64**, **double**, **bool** and **string**.
+**Null** is not allowed in matrix output.
+
+An exceptional case is  a **string** data type. In this case, the query result can contain different types. All data will be converted to a **string**.
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: extra space in "is  a"

Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: converted to a -> converted into

File: user_technical/reference_guide/tensorflow.md [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+A user must be careful here because converting data type to string, and vice versa can have unwanted performance issues.
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: no need for a comma

File: user_technical/reference_guide/tensorflow.md [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+
+#### Output lists:
+If the query contains list as output, the list expands into the row. All correspondent lists must have the same size.
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: https://en.wikipedia.org/wiki/Correspondent :)

**Comment by gitbuda** [comment]
* Remove whitespaces at the end of a line. That's the case in a lot of lines; please use some local tool to detect and/or remove the unnecessary whitespaces.
* All three are required eventually:
  1) How-to Guide: How-to setup TF Op?
  2) Tutorial: XYZ powered by Tensorflow and Memgraph
  3) Reference: Tensorflow Op
For now, it probably makes sense to split this reference guide into the reference guide and how-to guide. And then we can add a tutorial in the additional pull request.
* Please limit the length of each line to max 80 chars.

File: user_technical/reference_guide/tensorflow.md [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+All corresponding lists must have the same size.
+
+### Error Handling
+
+Memgraph TensorFlow op reports internal errors:
+
+  * Cannot create a client instance
+    * Memgraph TensorFlow op missing some system resources to
+    create a connection to Memgraph
+  * Cannot connect to memgraph: \<message\>
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: memgraph -> Memgraph

File: user_technical/how_to_guides/tensorflow-setup.md [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+
+```python3
+import tensorflow as tf
+
+# Load libmemgraph_op.so
+memgraph_op_module = tf.load_op_library('libmemgraph_op.so')
+
+
+def main():
+    query = """match (u :User)-->(m :Movie)
```
[the remark this answers is not in the document]
Comment (dsantl) [reply user_technical/how_to_guides/tensorflow-setup.md]: Multiple string lines.

File: user_technical/reference_guide/tensorflow.md [hunk 25]
```diff
@@ -25,41 +23,39 @@ There are two inputs:
   * query
   * input list
 
-The Query is a string. The query is an ```openCypher```
+The Query is a string. The query is an `openCypher`
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: Query -> query

File: user_technical/reference_guide/tensorflow.md [hunk 26]
```diff
@@ -25,41 +23,39 @@ There are two inputs:
...
+  * `host`, default: `127.0.0.1`
+  * `port`, default: `7687`
+  * `user`, default: `""` (empty string)
+  * `password`, default: `""` (empty string)
+  * `use_ssl`, default: `false`
+  * `output_dtype`
+
+`Host`, `port`, `user`, `password`
+and `use_ssl` are attributes used
+for connecting to Memgraph. The only different attribute is `output_type`.
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: Is this supposed to be `output_dtype`?

File: user_technical/reference_guide/tensorflow.md [hunk 27]
```diff
@@ -75,12 +71,12 @@ The list contains headers provided by query execution:
 MATCH (n) RETURN n.name as Name, n.address AS Address;
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: as -> AS

File: user_technical/reference_guide/tensorflow.md [hunk 28]
```diff
@@ -148,15 +144,15 @@ Matrix output:
 ### Limitations
 
 #### Input List
-Input list (```$input_list```) can contain only elements of **int64** type.
+Input list (`$input_list`) can contain only elements of **int64** type.
```
Comment (ipaljak) [inline user_technical/reference_guide/tensorflow.md]: **int64** -> `int64`

**Pull request merged**